File: ParseTiles.java
import java.nio.file.*;
import java.io.*;
import java.util.regex.*;

public class ParseTiles {
    public static void main(String[] args) throws Exception {
        String data = new String(Files.readAllBytes(Paths.get("c:\\Users\\agr_b\\MapRandomizer\\rust\\maprando\\src\\patch\\map_tiles.rs")));
        StringBuilder javaCode = new StringBuilder("package com.maprando.patch;\n\npublic class MapTilesArrays {\n");

        Pattern arrRegex = Pattern.compile("let (\\w+): \\[\\[u8; 8\\]; 8\\] = \\[(.*?)\\];", Pattern.DOTALL);
        Matcher m1 = arrRegex.matcher(data);
        while (m1.find()) {
            String name = m1.group(1);
            String content = m1.group(2).replace("[", "{").replace("]", "}");
            content = content.replaceAll("\\bb\\b", "15").replaceAll("\\bw\\b", "12");
            javaCode.append("    public static final int[][] ").append(name.toUpperCase()).append(" = {\n")
                    .append("        ").append(content.trim()).append("\n    };\n\n");
        }

        Pattern heatRegex = Pattern.compile("MapTileSpecialType::(\\w+) => \\{\\s*data = apply_heat\\(\\[\\s*(.*?)\\s*\\]\\);\\s*\\}", Pattern.DOTALL);
        Matcher m2 = heatRegex.matcher(data);
        while (m2.find()) {
            String name = m2.group(1).replaceAll("(.)([A-Z])", "$1_$2").toUpperCase();
            String content = m2.group(2).replace("[", "{").replace("]", "}");
            javaCode.append("    public static final int[][] ").append(name).append(" = {\n")
                    .append("        ").append(content.trim()).append("\n    };\n\n");
        }

        Pattern updRegex2 = Pattern.compile("update_tile\\(\\s*&mut data,\\s*\\d+,\\s*&vec!\\[(.*?)\\]\\s*,?\\s*\\);", Pattern.DOTALL);
        Matcher m3 = updRegex2.matcher(data);
        int i = 0;
        while (m3.find()) {
            String content = m3.group(1).replace("(", "{").replace(")", "}");
            javaCode.append("    public static final int[][] TILE_COORDS_").append(i).append(" = {\n")
                    .append("        ").append(content.trim()).append("\n    };\n\n");
            i++;
        }

        javaCode.append("}\n");

        Files.write(Paths.get("c:\\Users\\agr_b\\sm-rando-java\\src\\main\\java\\com\\maprando\\patch\\MapTilesArrays.java"), javaCode.toString().getBytes());
        System.out.println("Done extracting arrays!");
    }
}

File: parse_tiles.py
import re

def process_file():
    with open(r'c:\Users\agr_b\MapRandomizer\rust\maprando\src\patch\map_tiles.rs', 'r') as f:
        data = f.read()

    java_code = """package com.maprando.patch;

public class MapTilesArrays {
"""

    # Extract [8][8] variables
    arr_regex = re.compile(r'let (\w+): \[\[u8; 8\]; 8\] = \[(.*?)\];', re.DOTALL)
    for name, content in arr_regex.findall(data):
        content = content.replace('[', '{').replace(']', '}')
        # Replace b, w constants with 15, 12
        content = re.sub(r'\bb\b', '15', content)
        content = re.sub(r'\bw\b', '12', content)
        java_code += f"    public static final int[][] {name.upper()} = {{\n"
        java_code += f"        {content.strip()}\n"
        java_code += f"    }};\n\n"

    # Extract apply_heat data (Slopes)
    heat_regex = re.compile(r'MapTileSpecialType::(\w+) => \{\s*data = apply_heat\(\[\s*(.*?)\s*\]\);\s*\}', re.DOTALL)
    for name, content in heat_regex.findall(data):
        name_upper = re.sub(r'(?<!^)(?=[A-Z])', '_', name).upper()
        content = content.replace('[', '{').replace(']', '}')
        java_code += f"    public static final int[][] {name_upper} = {{\n"
        java_code += f"        {content.strip()}\n"
        java_code += f"    }};\n\n"

    # Save stations / Refill stations vector coordinates update_tile
    upd_regex = re.compile(r'MapTileInterior::(\w+) => \{\s*update_tile\(\s*&mut data,\s*\d+,\s*&vec!\[(.*?)\]\s*,?\s*\);\s*\}', re.DOTALL)
    for name, content in upd_regex.findall(data):
        name_upper = re.sub(r'(?<!^)(?=[A-Z])', '_', name).upper()
        content = content.replace('(', '{').replace(')', '}')
        java_code += f"    public static final int[][] {name_upper}_COORDS = {{\n"
        java_code += f"        {content.strip()}\n"
        java_code += f"    }};\n\n"

    # There's also some with if-else
    upd_regex2 = re.compile(r'update_tile\(\s*&mut data,\s*\d+,\s*&vec!\[(.*?)\]\s*,?\s*\);', re.DOTALL)
    i = 0
    for content in upd_regex2.findall(data):
        content = content.replace('(', '{').replace(')', '}')
        java_code += f"    public static final int[][] TILE_COORDS_{i} = {{\n"
        java_code += f"        {content.strip()}\n"
        java_code += f"    }};\n\n"
        i += 1

    java_code += "}\n"

    with open(r'c:\Users\agr_b\sm-rando-java\src\main\java\com\maprando\patch\MapTilesArrays.java', 'w') as f:
        f.write(java_code)

process_file()
